src/Microsoft.Framework.Runtime.Hosting/Json/JsonBuffer.cs

```diff
@@ -0,0 +1,312 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Globalization;
+using System.IO;
+using System.Text;
+
+namespace Microsoft.Framework.Runtime.Json
+{
+    internal class JsonBuffer
+    {
+        private readonly TextReader _reader;
+        private JsonToken _token;
+        private int _line;
+        private int _column;
+
+        public JsonBuffer(TextReader reader)
+        {
+            _reader = reader;
+            _line = 1;
+        }
+
+        public JsonToken Read()
+        {
+            int first;
+            while (true)
+            {
+                first = ReadNextChar();
+
+                if (first == -1)
+                {
+                    _token.Type = JsonTokenType.EOL;
+                    return _token;
+                }
+                else if (!IsWhitespace(first))
+                {
+                    break;
+                }
+            }
+
+            _token.Value = null;
+            _token.Line = _line;
+            _token.Column = _column;
+
+            if (first == '{')
```

Comment:
Here's an idea that I just had: create a dictionary that maps characters to tokens. If the dictionary contains that key, you take the value. It would reduce all this code to 3 lines :)